filters/openpolicyagent/response.go

```diff
@@ -0,0 +1,75 @@
+package openpolicyagent
+
+import (
+	"bytes"
+	"io"
+
+	"net/http"
+
+	"github.com/open-policy-agent/opa-envoy-plugin/envoyauth"
+	"github.com/opentracing/opentracing-go"
+	"github.com/zalando/skipper/filters"
+)
+
+func (opa *OpenPolicyAgentInstance) RejectInvalidDecisionError(fc filters.FilterContext, span opentracing.Span, result *envoyauth.EvalResult, err error) {
+	resp := http.Response{}
+
+	resp.StatusCode = http.StatusInternalServerError
+
+	if result != nil {
+		span.LogKV(
+			"event", "error",
+			"decision_id", result.DecisionID,
+			"message", err.Error(),
+		)
+
+		opa.Logger().WithFields(map[string]interface{}{
+			"decision":    result.Decision,
+			"err":         err,
+			"decision_id": result.DecisionID,
+		}).Info("Rejecting request because of an invalid decision")
+	} else {
+		span.LogKV(
+			"event", "error",
+			"message", err.Error(),
+		)
+
+		opa.Logger().WithFields(map[string]interface{}{
+			"err": err,
+		}).Info("Rejecting request because of an invalid decision")
+	}
+
+	fc.Serve(&resp)
+}
+
+func (opa *OpenPolicyAgentInstance) ServeResponse(fc filters.FilterContext, span opentracing.Span, result *envoyauth.EvalResult) {
+	resp := http.Response{}
+
+	status, err := result.GetResponseHTTPStatus()
```

Comment:
drop intermediate and use resp.StatusCode will reduce memory footprint:
```
var err error
resp.StatusCode , err =  result.GetResponseHTTPStatus()
if er != nil {...}
```

The same pattern applies below with `resp.Header`